### ISSUE TITLE
[TensorDict - compile] Custom `with stuff(param)` support in dynamo

### ISSUE THREAD

**Issue by vmoens** [pytorch/pytorch#129666]
### 🚀 The feature, motivation and pitch

Currently, a code [such as this will](https://github.com/pytorch/tensordict/blob/7252fcfa4719c1109187187b2290f71418b57ff1/test/test_compile.py#L404-L406) fail to be compiled.
```
...
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================= short test summary info =======================================================================
FAILED test/test_compile.py::TestFunctional::test_functional[False] - torch._dynamo.exc.Unsupported: BEFORE_WITH UserDefinedObjectVariable(TensorDict)
```
The alternative code (just below, uncommented) works fine.

cc @ezyang @anijain2305 @chauhang @penguinwu

**Comment by anijain2305** [comment]
https://github.com/pytorch/pytorch/pull/129091 is a pre-req for this. I tried that PR but the test still fails. So, I think there is more work required. I will revisit this once the PR is merged.

**Comment by vmoens** [comment]
Closing in favor of #141964
